### ISSUE TITLE
ttx dies on file with no glyphs

### ISSUE THREAD

**Issue by simoncozens** [fonttools/fonttools#669]
I created a new file in Glyphs, turned off exporting for all characters, and exported as OTF. Then I fed it to `ttx`:

```
$ ttx TTXTest-Regular.otf
Dumping "TTXTest-Regular.otf" to "TTXTest-Regular.ttx"...
Dumping 'GlyphOrder' table...
Dumping 'head' table...
Dumping 'hhea' table...
Dumping 'maxp' table...
Dumping 'OS/2' table...
Dumping 'name' table...
Dumping 'cmap' table...
Dumping 'post' table...
Dumping 'CFF ' table...
.notdef
A
ERROR: Unhandled exception has occurred
Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/FontTools/fontTools/ttx.py", line 383, in main
    process(jobs, options)
  File "/Library/Python/2.7/site-packages/FontTools/fontTools/ttx.py", line 356, in process
    action(input, output, options)
  File "/Library/Python/2.7/site-packages/FontTools/fontTools/misc/loggingTools.py", line 372, in wrapper
    return func(*args, **kwds)
  File "/Library/Python/2.7/site-packages/FontTools/fontTools/ttx.py", line 256, in ttDump
    bitmapGlyphDataFormat=options.bitmapGlyphDataFormat)
  File "/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/__init__.py", line 302, in saveXML
    self._tableToXML(tableWriter, tag, progress)
  File "/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/__init__.py", line 338, in _tableToXML
    table.toXML(writer, self, progress)
  File "/Library/Python/2.7/site-packages/FontTools/fontTools/ttLib/tables/C_F_F_.py", line 42, in toXML
    self.cff.toXML(writer, progress)
  File "/Library/Python/2.7/site-packages/FontTools/fontTools/cffLib.py", line 86, in toXML
    font.toXML(xmlWriter, progress)
  File "/Library/Python/2.7/site-packages/FontTools/fontTools/cffLib.py", line 1549, in toXML
    BaseDict.toXML(self, xmlWriter, progress)
  File "/Library/Python/2.7/site-packages/FontTools/fontTools/cffLib.py", line 1508, in toXML
    conv.xmlWrite(xmlWriter, name, value, progress)
  File "/Library/Python/2.7/site-packages/FontTools/fontTools/cffLib.py", line 720, in xmlWrite
    value.toXML(xmlWriter, progress)
  File "/Library/Python/2.7/site-packages/FontTools/fontTools/cffLib.py", line 564, in toXML
    charStr, fdSelectIndex = self.getItemAndSelector(name)
  File "/Library/Python/2.7/site-packages/FontTools/fontTools/cffLib.py", line 546, in getItemAndSelector
    return self.charStringsIndex.getItemAndSelector(index)
  File "/Library/Python/2.7/site-packages/FontTools/fontTools/cffLib.py", line 407, in getItemAndSelector
    return self[index], sel
  File "/Library/Python/2.7/site-packages/FontTools/fontTools/cffLib.py", line 337, in __getitem__
    item = self.items[index]
IndexError: list index out of range
```

The `.notdef` and `A` lines in there are a debugging print which I inserted into `cffLib.py`:

    def toXML(self, xmlWriter, progress):
        names = sorted(self.keys())
        i = 0
        step = 10
        numGlyphs = len(names)
        for name in names:
            print(name)

So it sees that something in the font is claiming to have an `A` glyph, but there isn't one defined. I don't know if Glyphs is producing broken OTFs (fontlint doesn't think so) but you're certainly missing an index check somewhere.

**Comment by anthrotype** [comment]
Could you please attach the font here so we can try to reproduce?
Thank you.

**Comment by simoncozens** [comment]
Here you go!

[ttx-669.zip](https://github.com/behdad/fonttools/files/481727/ttx-669.zip)

**Comment by khaledhosny** [comment]
I think the CFF table is broken. It uses a predefined charset (id=0, ISOAdobe) and the spec says:
>  A font may use a predefined charset if it exactly matches in the first nGlyphs

But the font has exactly one glyph, so it does not match the predefined charset.

**Comment by simoncozens** [comment]
Which makes this a... Glyphs bug? makeotf bug?

**Comment by khaledhosny** [comment]
Whatever created the font.

**Comment by khaledhosny** [comment]
Anyway, here is a simple patch to work around this, not sure if it is a good idea though:
```patch
diff --git a/Lib/fontTools/cffLib.py b/Lib/fontTools/cffLib.py
index f038f2a..979618f 100644
--- a/Lib/fontTools/cffLib.py
+++ b/Lib/fontTools/cffLib.py
@@ -806,6 +806,11 @@ class CharsetConverter(object):
                                charset = cffIExpertStrings
                        elif value == 2:
                                charset = cffExpertSubsetStrings
+
+                       # Workaround buggy fonts
+                       # see https://github.com/behdad/fonttools/issues/669
+                       if len(charset) != parent.numGlyphs:
+                               charset = charset[:parent.numGlyphs]
                return charset
 
        def write(self, parent, value):
```

**Comment by simoncozens** [comment]
Well, I would personally check the length before indexing the array. But either way I think it's good defensive programming. You don't want third-party untrusted data to be crashing your library.

And I'll take it up with the Glyphs side.

**Comment by justvanrossum** [comment]
"FontTools, exposing bugs in other software since 1999." :)

**Comment by khaledhosny** [comment]
I’d just assert and call it a day, so many things can be wrong here.

**Comment by schriftgestalt** [comment]
I had a look at this. Glyphs is exporting the \Encoding since the beginning. Nobody complained about this, yet. Is the Encoding in the cff table used for anything? The names are stored independently. If I add a custom encoding, it just duplicates the glyph names.

**Comment by adrientetar** [comment]
> Is the Encoding in the cff table used for anything?

I think it's mostly an artifact of CFF as standalone format and is removed in CFF2.

**Comment by schriftgestalt** [comment]
So the code was using the encoding to then access the "items". Why not use the items directly?

**Comment by khaledhosny** [comment]
FontTools seems to use CFF charset for GlyphOrder. May be there is a alternative way to do this, but given that we don’t have many reports of similarly failing fonts I’m not sure if it is worth trying to find one.

**Comment by anthrotype** [comment]
What such an alternative way could be? The `post` table in CFF does not contain the glyph names.

**Comment by schriftgestalt** [comment]
Now it loads the encoding as a basis for the glyph order. And in case of the Standard encoding, it loads the list and the iterated through it to access the charstring. I would suggest to not use the encoding but read the charstrings directly.

**Comment by anthrotype** [comment]
Perhaps @readroberts could shed some light on this.

**Comment by readroberts** [comment]
I have several comments. 1) The code is buggy, not the font. I quote from the CFF spec: "A font may use a predefined charset if it exactly matches in the first nGlyphs". ".notdef" is, by definition part of all all charsets and is omitted from the charset glyph list, so the glyph list from this font actually would match any of the predefined charsets. 2) A different fix is probably better: I haven't checked, but I suspect that this charset will be written back out as a custom charset with 0 glyph names. I am looking at cffLib.py for other reasons, and will now look into this as well. 3) The "Encoding" field is useful and important only when the CFF data is being used as a stand-alone resource. When the CFF data exists as an OpenType font table, the "Encoding field" should be ignored (along with a lot of other CFF data which is duplicated in other OpenType tables). Adobe recommends always setting it to StandardEncoding in this case. 4) 'charset' and 'Encoding' are different entities. 'Encoding' is the Type 1 encoding, for mapping a Postscript text string character code to a glyph id, and has no correlation with glyph names or order in the font. 'charset' provides the only definition of  glyph names. It does not define glyph order: instead, the order of the glyph names in the 'charset' is defined by the order of the glyphs in the  CharStrings INDEX.

**Comment by readroberts** [comment]
Well, I was wrong. I assumed that fonts with a glyph name order that matched an initial subset of the predefined charsets were working other than the current problem case, but this is not so. I guess there must be few fonts out there where both glyph names and  glyph order match an initial subset of a  predefined charset, or this would have been reported before. The fix above actually is correct, and is the right place to do it. cffLib also needs an optimization when compiling from XML: right now, it never checks to see if it can use any of the predefined charset values.

**Comment by simoncozens** [comment]
"@simoncozens, exposing ambiguities in software specifications since 1999."

**Comment by anthrotype** [comment]
I believe this should be fixed with 99394d90bb96f0b211e1d73885c5a884ca447f5d.
Please reopen if that's not the case.
